### PR TITLE
Rotation gadget for Snarky

### PR DESCRIPTION
**Pull request by querolita**
This PR creates the rotation gadget for 64-bits. 

Closes https://github.com/MinaProtocol/mina/issues/12884

### CODE REVIEW

**Comment by querolita** [comment]
!ci-build-me

**Comment by querolita** [comment]
!ci-build-me

**Comment by querolita** [comment]
!ci-build-me

File: src/lib/crypto/kimchi_backend/gadgets/rot.ml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  let values =
+    exists (Typ.array ~length:4 Field.typ) ~compute:(fun () ->
+        (* Assert that word is at most 64 bits*)
+        let word_big =
+          Common.(
+            field_to_bignum_bigint
+              (module Circuit)
+              (cvar_field_to_field_as_prover (module Circuit) word))
+        in
+        assert (Bignum_bigint.(word_big < big_2_pow_64)) ;
```
Comment (jspada) [inline src/lib/crypto/kimchi_backend/gadgets/rot.ml:40]: nice!

File: src/lib/crypto/kimchi_backend/gadgets/rot.ml [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+             word * 2^rot = excess * 2^64 + shifted *)
+        let excess_big, shifted_big =
+          Common.bignum_bigint_div_rem
+            Bignum_bigint.(word_big * big_2_pow_rot)
+            big_2_pow_64
+        in
+
+        (* Compute rotated value as
+           rotated = excess + shifted *)
+        let rotated_big = Bignum_bigint.(shifted_big + excess_big) in
```
Comment (jspada) [inline src/lib/crypto/kimchi_backend/gadgets/rot.ml:52]: great commenting.  makes it clear to read

File: src/lib/crypto/kimchi_backend/gadgets/rot.ml [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+        (* Compute bound that is the right input of FFAdd equation *)
+        let bound_big =
+          Bignum_bigint.(excess_big + big_2_pow_64 - big_2_pow_rot)
```
Comment (jspada) [inline src/lib/crypto/kimchi_backend/gadgets/rot.ml:56]: Double check this is correct.  Isn't it `2^64 - 2^rot` as it is in the rust version?

File: src/lib/crypto/kimchi_backend/gadgets/rot.ml [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  in
+
+  let of_bits =
+    Common.as_prover_cvar_field_bits_le_to_cvar_field (module Circuit)
+  in
+
+  let rotated = values.(0) in
+  let excess = values.(1) in
+  let shifted = values.(2) in
+  let bound = values.(3) in
```
Comment (jspada) [inline src/lib/crypto/kimchi_backend/gadgets/rot.ml:81]: lgtm

File: src/lib/crypto/kimchi_backend/gadgets/rot.ml [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                 ; bound_limb1 = of_bits bound 40 52
+                 ; bound_limb2 = of_bits bound 28 40
+                 ; bound_limb3 = of_bits bound 16 28
+                 ; bound_crumb0 = of_bits bound 14 16
+                 ; bound_crumb1 = of_bits bound 12 14
+                 ; bound_crumb2 = of_bits bound 10 12
+                 ; bound_crumb3 = of_bits bound 8 10
+                 ; bound_crumb4 = of_bits bound 6 8
+                 ; bound_crumb5 = of_bits bound 4 6
+                 ; bound_crumb6 = of_bits bound 2 4
```
Comment (jspada) [inline src/lib/crypto/kimchi_backend/gadgets/rot.ml:104]: these lgtm!

File: src/lib/crypto/kimchi_backend/gadgets/rot.ml [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+open Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint
+
+module Bignum_bigint = Snarky_backendless.Backend_extended.Bignum_bigint
+
+(* ROT *)
+
+(* Side of rotation *)
+type rot_mode = Left | Right
+
+(* 64-bit rotation of rot_bits to the `mode` side *)
```
Comment (jspada) [inline src/lib/crypto/kimchi_backend/gadgets/rot.ml:12]: Comment inputs and outputs to gadget.

File: src/lib/crypto/kimchi_backend/gadgets/rot.ml [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  test_rot "1" 63 Left "9223372036854775808" ;
+  test_rot "256" 4 Right "16" ;
+  test_rot "1234567890" 32 Right "5302428712241725440" ;
+  (* 0x5A5A5A5A5A5A5A5A is 0xA5A5A5A5A5A5A5A5 both when rotate 4 bits Left or Right*)
+  test_rot "6510615555426900570" 4 Left "11936128518282651045" ;
+  test_rot "6510615555426900570" 4 Right "11936128518282651045" ;
+
+  (* Negatve tests *)
+  assert (Common.is_error (fun () -> test_rot "0" 1 Left "1")) ;
+  assert (Common.is_error (fun () -> test_rot "1" 64 Left "1")) ;
```
Comment (jspada) [inline src/lib/crypto/kimchi_backend/gadgets/rot.ml:163]: Perhaps there's another interesting negative test case you can think of.

**Review by jspada — APPROVED** [review]
Approved with some minor suggestions, but one important question.

File: src/lib/crypto/kimchi_backend/gadgets/rot.ml (hunk 3 above)
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/rot.ml:56]: No because in the rust version the bound operation is performed later in the witness code

**Pull request merged**